File: android/src/main/java/ru/reactnativerustorepush/deps/MessagingService.kt
package ru.reactnativerustorepush.deps

import android.content.Intent
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.encodeToString

import ru.rustore.sdk.pushclient.messaging.model.Notification
import ru.rustore.sdk.pushclient.messaging.model.RemoteMessage
import ru.rustore.sdk.pushclient.messaging.service.RuStoreMessagingService
import ru.rustore.sdk.pushclient.messaging.exception.RuStorePushClientException
import ru.reactnativerustorepush.deps.Constants
import ru.reactnativerustorepush.deps.RuStorePushModuleEvent

@Serializable
data class NewTokenEventPayload(
    val token: String,
)

@Serializable
data class NotificationObject(
    val title: String?,
    val body: String?,
    val channelId: String?,
    val imageUrl: String?,
    val color: String?,
    val icon: String?, 
    val clickAction: String?,
)

@Serializable
data class MessageObject(
    val messageId: String?,
    val priority: Int,
    val ttl: Int,
    val collapseKey: String?,
    val data: Map<String, String>,
    val rawData: ByteArray?,
    val notification: NotificationObject?,
)

class MessagingService : RuStoreMessagingService() {

    fun log(eventName: String, eventPayload: String) {
        val intent = Intent(Constants.MESSAGING_SERVICE_TAG)
        val msg = Json.encodeToString(RuStorePushModuleEvent(eventName, eventPayload))

        intent.putExtra(Constants.EXTRA_FIELD_NAME, msg)

        sendBroadcast(intent)
    }

    override fun onNewToken(token: String) {
        val eventPayload = Json.encodeToString(NewTokenEventPayload(token))

        log("new-token", eventPayload)
    }

    override fun onMessageReceived(message: RemoteMessage) {
        var notification: NotificationObject? = null

        if (message.notification != null) {
            val messageNotification = message.notification as Notification

            notification = NotificationObject(
                title = messageNotification.title,
                body = messageNotification.body,
                channelId = messageNotification.channelId,
                imageUrl = messageNotification.imageUrl?.toString(),
                color = messageNotification.color,
                icon = messageNotification.icon,
                clickAction = messageNotification.clickAction,
            )
        }

        val eventPayload = Json.encodeToString(
            MessageObject(
                messageId = message.messageId,
                priority = message.priority,
                ttl = message.ttl,
                collapseKey = message.collapseKey,
                data = message.data,
                rawData = message.rawData,
                notification = notification,
            )
        )

        log("message-received", eventPayload)
    }

    override fun onDeletedMessages() {
		log("deleted-messages", "\"null\"")
    }

    override fun onError(errors: List<RuStorePushClientException>) {
        val list: MutableList<String> = mutableListOf()

        errors.forEach { error -> list.add(error.toString()) }

        log("error", Json.encodeToString(list))
    }
}

File: src/package/messaging-service.ts
import { NativeEventEmitter } from 'react-native';

import {
	RuStorePushModule, RuStorePushNativeModule, parseEventPayload, parseEvent,
} from './native-module';

/** Приоритет пуш-уведомления */
enum Priority {
	UNKNOWN,
	HIGH,
	NORMAL,
}

/** Пуш-токен */
type Token = string;

type Notification = {
	/** Заголовок уведомления */
    title?: string
	/** Тело уведомления */
    body?: string
	/** Канал, в который будет отправлено уведомление (актуально для Android 8.0 и выше) */
    channelId?: string
	/** Прямая ссылка на изображение для вставки в уведомление (изображение должно быть не более 1 мегабайта) */
    imageUrl?: string
	/** Цвет уведомления. Цвет необходимо передать в hex-формате (например `"#A52A2A"`) */
    color?: string
	/**
	 * Иконка уведомления. Иконка должна лежать в ресурсах приложения (`res/drawable`).
	 *
	 * Пример: в `res/drawable` лежит иконка `small_icon.xml`.
	 * Для отображения данной иконки в уведомлении сервер должен поместить в параметр `icon` значение `"small_icon"`.
	 */
    icon?: string
	/** Intent action, с помощью которого будет открыта активити при нажатии на уведомление (не поддерживается данным пакетом) */
    clickAction?: string
}

type RemoteMessage = {
	/** Уникальный ID сообщения. Является идентификатором каждого сообщения. */
    messageId?: string
	/** Значение приоритетности (на данный момент не учитывается) */
    priority: Priority
	/** Время жизни пуш-уведомления в секундах */
    ttl: number
	/** Идентификатор группы уведомлений (на данный момент не учитывается) */
    collapseKey?: string
	/** Объект с дополнительными данными */
    data: Record<string, string>
	/** Содержимое объекта `data` в бинарном виде  */
    rawData?: ArrayBuffer
	/** Объект уведомления */
    notification?: Notification
}

type EventListeners = {
	'new-token': Token
	'message-received': RemoteMessage
	'deleted-messages': null
	'error': string[]
};

type EventName = keyof EventListeners;

type EventListenerObject<Name extends EventName> = {
	name: Name
	handler(data: EventListeners[Name]): void
};

type UnsubscribeFunction = () => void;

class MessagingService {
	private listeners: Array<EventListenerObject<EventName>>;

	private ee: NativeEventEmitter;

	constructor() {
		this.listeners = [];
		this.ee = new NativeEventEmitter(RuStorePushNativeModule);

		this.init();
	}

	private async init() {
		const constants = await RuStorePushModule.getConstants();

		this.ee.addListener(constants.MESSAGING_SERVICE_TAG, (event: string) => {
			const parsedEvent = parseEvent(event);

			if (parsedEvent) {
				this.listeners.forEach((listener) => {
					const { name, handler } = listener;

					if (parsedEvent.name === name) {
						const payload = parseEventPayload<EventListeners[typeof name]>(parsedEvent.payload);

						if (payload !== undefined) {
							handler(payload);
						}
					}
				});
			}
		});
	}

	/**
	 * Добавить обработчик на событие получения данных от **RuStore SDK**.
	 *
	 * Имеется четыре вида обработчиков событий:
	 *
	 * 1. **`new-token`** - будет вызван при получении нового пуш-токена.
	 * После вызова данного метода ваше приложение ответственно за то, чтобы передать новый пуш-токен на свой серве.
	 *
	 * 2. **`message-received`** - будет вызван при получении нового пуш-уведомления.
	 * Если в объекте `notification` есть данные, то **RuStore SDK** сама отобразит уведомление.
	 * Если вы не хотите, чтобы **RuStore SDK** самостоятельно отображала уведомление,
	 * то используйте объект `data`, а объект `notification` оставляйте пустым.
	 * Однако сам обработчик будет вызван в любом случае.
	 * Получить полезную нагрузку пуш-уведомления можно из поля `data`.
	 *
	 * 3. **`deleted-messages`** - будет вызван в случае, если одно или несколько пуш-уведомлений не были доставлены на устройство.
	 * Это может произойти, например, по причине истечения времени жизни уведомления до того, как оно будет доставлено на устройство.
	 * При вызове этого метода рекомендуется синхронизироваться со своим сервером, чтобы не пропустить данные.
	 *
	 * 4. **`error`** - будет вызван при возникновении ошибки в момент инициализации.
	 * @param event Имя события
	 * @param listener Обработчик
	 * @returns Функция для уничтожения созданного обработчика
	 */
	on<Event extends EventName>(event: Event, listener: (data: EventListeners[Event]) => void) {
		const listenerObj: EventListenerObject<Event> = {
			name: event,
			handler: listener,
		};

		this.listeners.push(listenerObj);

		const unsubscribe: UnsubscribeFunction = () => {
			this.listeners = this.listeners.filter((el) => el !== listenerObj);
		};

		return unsubscribe;
	}

	/** Уничтожить все обработчики событий. */
	unsubscribeAll() {
		this.listeners = [];
	}
}

export default MessagingService;
